File: pnetlib/runtime/System/Reflection/Module.cs
/*
 * Module.cs - Implementation of the "System.Reflection.Module" class.
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

namespace System.Reflection
{

#if CONFIG_REFLECTION

using System;
using System.Collections;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Runtime.Serialization;
using System.Security.Cryptography.X509Certificates;

public class Module : IClrProgramItem, ICustomAttributeProvider
#if CONFIG_SERIALIZATION
	, ISerializable
#endif
{
	// Internal state.
	internal IntPtr privateData;

	// Constructor.
	internal Module() {}

	// Implement the IClrProgramItem interface.
	IntPtr IClrProgramItem.ClrHandle
			{
				get
				{
					return privateData;
				}
			}

	// Get a field from the module type.
	public FieldInfo GetField(String name)
			{
				return GetModuleType().GetField(name);
			}
	public FieldInfo GetField(String name, BindingFlags bindingAttr)
			{
				return GetModuleType().GetField(name, bindingAttr);
			}

	// Get all module type fields that match a specific set of criteria.
	// Note: the first is required by ECMA, but doesn't exist in the SDK,
	// so you should avoid using it if at all possible.
	public FieldInfo[] GetFields(BindingFlags bindingAttr)
			{
				return GetModuleType().GetFields(bindingAttr);
			}
	public FieldInfo[] GetFields()
			{
				return GetModuleType().GetFields();
			}

	// Get a method from the module type.
	public MethodInfo GetMethod(String name)
			{
				return GetMethodImpl(name, BindingFlags.Default,
									 null, CallingConventions.Any,
									 new Type [0], null);
			}
	public MethodInfo GetMethod(String name, Type[] types)
			{
				return GetMethodImpl(name, BindingFlags.Default,
									 null, CallingConventions.Any,
									 types, null);
			}
#if !ECMA_COMPAT
	public MethodInfo GetMethod(String name,
							    BindingFlags bindingAttr, Binder binder,
								CallingConventions callingConventions,
								Type[] types, ParameterModifier[] modifiers)
			{
				return GetMethodImpl(name, bindingAttr, binder,
									 callingConventions, types, modifiers);
			}
#endif // !ECMA_COMPAT

	// Get a method implementation that matches a specific set of criteria.
#if ECMA_COMPAT
	internal
#endif
	protected virtual MethodInfo GetMethodImpl
				(String name, BindingFlags bindingAttr, Binder binder,
				 CallingConventions callingConventions,
				 Type[] types, ParameterModifier[] modifiers)
			{
				return GetModuleType().GetMethod
						(name, bindingAttr, binder,
						 callingConventions, types, modifiers);
			}

	// Get all module type methods that match a specific set of criteria.
	// Note: the first is required by ECMA, but doesn't exist in the SDK,
	// so you should avoid using it if at all possible.
	public MethodInfo[] GetMethods(BindingFlags bindingAttr)
			{
				return GetModuleType().GetMethods(bindingAttr);
			}
	public MethodInfo[] GetMethods()
			{
				return GetModuleType().GetMethods();
			}

	// Convert the module into a string.
	public override String ToString()
			{
				return Name;
			}

#if CONFIG_RUNTIME_INFRA
	// Get the assembly that contains this module.
	public System.Reflection.Assembly Assembly
			{
				get
				{
					return GetAssemblyCore();
				}
			}

	// Workaround the fact that the Assembly property is non-virtual.
	internal virtual System.Reflection.Assembly GetAssemblyCore()
			{
				return GetAssembly();
			}
#endif

	// Get the fully-qualified name for this module.
	public virtual String FullyQualifiedName
			{
				get
				{
					return GetFullName();
				}
			}

	// Get the name of this module.
	public String Name
			{
				get
				{
					return ClrHelpers.GetName(privateData);
				}
			}

#if !ECMA_COMPAT

	// Get the scope name for this module.
	public String ScopeName
			{
				get
				{
					return ClrHelpers.GetName(privateData);
				}
			}

	// Get custom attributes for this module.
	public virtual Object[] GetCustomAttributes(bool inherit)
			{
				return ClrHelpers.GetCustomAttributes(this, inherit);
			}
	public virtual Object[] GetCustomAttributes(Type type, bool inherit)
			{
				return ClrHelpers.GetCustomAttributes(this, type, inherit);
			}
	
	// Determine if a custom attribute is defined for this module.
	public virtual bool IsDefined(Type type, bool inherit)
			{
				return ClrHelpers.IsDefined(this, type, inherit);
			}

	// Get a type from this module.
	public virtual Type GetType(String name)
			{
				return GetType(name, false, false);
			}
	public virtual Type GetType(String name, bool ignoreCase)
			{
				return GetType(name, false, ignoreCase);
			}
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern public virtual Type GetType(String name, bool throwOnError,
									   bool ignoreCase);

	// Return all types that are defined in this module.
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern public virtual Type[] GetTypes();

	// Determine if this module is a resource.
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern public bool IsResource();

#if CONFIG_X509_CERTIFICATES

	// Get the signer certificate for this module.
	public X509Certificate GetSignerCertificate()
			{
				// Not used in this implementation.
				return null;
			}

#endif

#if CONFIG_SERIALIZATION

	// Get the serialization data for this module.
	public virtual void GetObjectData(SerializationInfo info,
							  		  StreamingContext context)
			{
				if(info == null)
				{
					throw new ArgumentNullException("info");
				}
				UnitySerializationHolder.Serialize
					(info, UnitySerializationHolder.UnityType.Module,
					 ScopeName, Assembly);
			}

#endif // CONFIG_SERIALIZATION

	// A type filter that searches on name.
	private static bool FilterTypeNameImpl(Type type, Object criteria)
			{
				String name = (criteria as String);
				if(name != null)
				{
					if(name.EndsWith("*"))
					{
						return type.FullName.StartsWith
							(name.Substring(0, name.Length - 1));
					}
					else
					{
						return (type.FullName == name);
					}
				}
				return false;
			}

	// A type filter that searches on name while ignoring case.
	private static bool FilterTypeNameIgnoreCaseImpl
				(Type type, Object criteria)
			{
				String name = (criteria as String);
				if(name != null)
				{
					if(name.EndsWith("*"))
					{
						name = name.Substring(0, name.Length - 1);
						String fullName = type.FullName;
						if(fullName.Length < name.Length)
						{
							return false;
						}
						return (String.Compare(fullName, 0, name, 0,
											   name.Length, true) == 0);
					}
					else
					{
						return (String.Compare(type.FullName, name, true)
									== 0);
					}
				}
				return false;
			}

	// Declare the standard type filters.
	public static readonly TypeFilter FilterTypeName =
			new TypeFilter(FilterTypeNameImpl);
	public static readonly TypeFilter FilterTypeNameIgnoreCase =
			new TypeFilter(FilterTypeNameIgnoreCaseImpl);

	// Find all types that match a specified filter criteria.
	public virtual Type[] FindTypes(TypeFilter filter,
									Object filterCriteria)
			{
				Type[] types = GetTypes();
				if(filter == null || types == null)
				{
					return types;
				}
				ArrayList list = new ArrayList();
				foreach(Type type in types)
				{
					if(filter(type, filterCriteria))
					{
						list.Add(type);
					}
				}
				return (Type[])(list.ToArray(typeof(Type)));
			}

#else // ECMA_COMPAT

	// Get custom attributes for this module.
	Object[] ICustomAttributeProvider.GetCustomAttributes(bool inherit)
			{
				return ClrHelpers.GetCustomAttributes(this, inherit);
			}
	Object[] ICustomAttributeProvider.GetCustomAttributes
					(Type type, bool inherit)
			{
				return ClrHelpers.GetCustomAttributes(this, type, inherit);
			}
	
	// Determine if a custom attribute is defined for this module.
	bool ICustomAttributeProvider.IsDefined(Type type, bool inherit)
			{
				return ClrHelpers.IsDefined(this, type, inherit);
			}

#endif // ECMA_COMPAT

	// Get the module type.
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern internal virtual Type GetModuleType();

	// Get the assembly that contains this module.
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern private Assembly GetAssembly();

	// Get the fully-qualified name for this module.
	[MethodImpl(MethodImplOptions.InternalCall)]
	extern private String GetFullName();

}; // class Module

#endif // CONFIG_REFLECTION

}; // namespace System.Reflection

File: pnet/image/module.c
/*
 * module.c - Process module information from an image file.
 *
 * Copyright (C) 2001  Southern Storm Software, Pty Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "program.h"

#ifdef	__cplusplus
extern	"C" {
#endif

ILModule *ILModuleCreate(ILImage *image, ILToken token,
						 const char *name, const unsigned char *mvid)
{
	ILModule *module;

	/* Allocate space for the module from the memory stack */
	module = ILMemStackAlloc(&(image->memStack), ILModule);
	if(!module)
	{
		return 0;
	}
	module->programItem.image = image;

	/* Convert the module name into a persistent string */
	if(name)
	{
		module->name = _ILContextPersistString(image, name);
		if(!(module->name))
		{
			return 0;
		}
	}

	/* Assign a token code to the module */
	if(!_ILImageSetToken(image, &(module->programItem),
						 token, IL_META_TOKEN_MODULE))
	{
		return 0;
	}

	/* Create the MVID field for the module */
	module->mvid = (unsigned char *)ILMemStackAllocItem(&(image->memStack), 16);
	if(!(module->mvid))
	{
		return 0;
	}
	if(mvid)
	{
		ILMemCpy(module->mvid, mvid, 16);
	}
	else
	{
		ILGUIDGenerate(module->mvid);
	}

	/* Return the module to the caller */
	return module;
}

int ILModuleSetName(ILModule *module, const char *name)
{
	module->name = _ILContextPersistString(module->programItem.image, name);
	return (module->name != 0);
}

const char *ILModuleGetName(ILModule *module)
{
	return module->name;
}

void ILModuleSetMVID(ILModule *module, const unsigned char *mvid)
{
	/* If "module->mvid" is NULL, then this is a module reference
	   that does not have an MVID associated with it */
	if(module->mvid)
	{
		if(mvid)
		{
			ILMemCpy(module->mvid, mvid, 16);
		}
		else
		{
			ILGUIDGenerate(module->mvid);
		}
	}
}

const unsigned char *ILModuleGetMVID(ILModule *module)
{
	return module->mvid;
}

/*
 * Create the Edit & Continue information for a module.
 */
static ILModuleEnc *CreateEnc(ILModule *module)
{
	if(module->enc)
	{
		/* We already have Edit & Continue information */
		return module->enc;
	}
	else
	{
		/* Create a new Edit & Continue block */
		module->enc = ILMemStackAlloc(&(module->programItem.image->memStack),
									  ILModuleEnc);
		return module->enc;
	}
}

int ILModuleSetGeneration(ILModule *module, ILUInt32 generation)
{
	ILModuleEnc *enc = CreateEnc(module);
	if(enc)
	{
		enc->generation = generation;
		return 1;
	}
	else
	{
		return 0;
	}
}

ILUInt32 ILModuleGetGeneration(ILModule *module)
{
	if(module->enc)
	{
		return module->enc->generation;
	}
	else
	{
		return 0;
	}
}

int ILModuleSetEncId(ILModule *module, const unsigned char *id)
{
	ILModuleEnc *enc = CreateEnc(module);
	if(enc)
	{
		ILMemCpy(enc->encId, id, 16);
		enc->flags |= 1;
		return 1;
	}
	else
	{
		return 0;
	}
}

const unsigned char *ILModuleGetEncId(ILModule *module)
{
	if(module->enc && (module->enc->flags & 1) != 0)
	{
		return module->enc->encId;
	}
	else
	{
		return 0;
	}
}

int ILModuleSetEncBaseId(ILModule *module, const unsigned char *id)
{
	ILModuleEnc *enc = CreateEnc(module);
	if(enc)
	{
		ILMemCpy(enc->encBaseId, id, 16);
		enc->flags |= 2;
		return 1;
	}
	else
	{
		return 0;
	}
}

const unsigned char *ILModuleGetEncBaseId(ILModule *module)
{
	if(module->enc && (module->enc->flags & 2) != 0)
	{
		return module->enc->encBaseId;
	}
	else
	{
		return 0;
	}
}

ILModule *ILModuleRefCreate(ILImage *image, ILToken token, const char *name)
{
	ILModule *module;

	/* Allocate space for the module reference from the memory stack */
	module = ILMemStackAlloc(&(image->memStack), ILModule);
	if(!module)
	{
		return 0;
	}
	module->programItem.image = image;

	/* Convert the module name into a persistent string */
	if(name)
	{
		module->name = _ILContextPersistString(image, name);
		if(!(module->name))
		{
			return 0;
		}
	}

	/* Assign a token code to the module */
	if(!_ILImageSetToken(image, &(module->programItem),
						 token, IL_META_TOKEN_MODULE_REF))
	{
		return 0;
	}

	/* Return the module reference to the caller */
	return module;
}

ILModule *ILModuleRefCreateUnique(ILImage *image, const char *name)
{
	ILModule *module;

	/* Search for an existing reference with the given name */
	module = 0;
	while((module = (ILModule *)ILImageNextToken
					(image, IL_META_TOKEN_MODULE_REF, module)) != 0)
	{
		if(module->name && !ILStrICmp(module->name, name))
		{
			return module;
		}
	}

	/* Create a new reference with the name */
	return ILModuleRefCreate(image, 0, name);
}

int ILModuleIsRef(ILModule *module)
{
	return ((module->programItem.token & IL_META_TOKEN_MASK) ==
					IL_META_TOKEN_MODULE_REF);
}

ILImage *ILModuleToImage(ILModule *module)
{
	if(!ILModuleIsRef(module))
	{
		/* Refers to the current image */
		return module->programItem.image;
	}
	else if(module->programItem.linked)
	{
		/* We have cached a previous image */
		return _ILProgramItemResolve(&(module->programItem))->image;
	}
	else
	{
		/* Search for a matching ILModule in another image */
		ILImage *image;
		ILModule *newModule = 0;
		image = module->programItem.image->context->firstImage;
		while(image != 0 && newModule == 0)
		{
			newModule = 0;
			while((newModule = (ILModule *)ILImageNextToken
						(image, IL_META_TOKEN_MODULE, newModule)) != 0)
			{
				if(!ILStrICmp(newModule->name, module->name))
				{
					break;
				}
			}
			image = image->nextImage;
		}
		if(!newModule)
		{
			return 0;
		}

		/* Link the reference to the foreign module */
		_ILProgramItemLink(&(module->programItem), &(newModule->programItem));

		/* Return the new module's image to the caller */
		return newModule->programItem.image;
	}
}

#ifdef	__cplusplus
};
#endif
